### PR TITLE
Add automatic cleanup on Singleton removal

### PR DESCRIPTION
**Pull request by Nokel81**
Signed-off-by: Sebastian Malton <sebastian@malton.name>

### CODE REVIEW

File: src/main/window-manager.ts [hunk 1]
```diff
@@ -214,9 +206,5 @@ export class WindowManager extends Singleton {
     this.splashWindow.destroy();
     this.mainWindow = null;
     this.splashWindow = null;
-    Object.entries(this.disposers).forEach(([name, dispose]) => {
-      dispose();
-      delete this.disposers[name];
-    });
```
Comment (jim-docker) [inline src/main/window-manager.ts:221]: Should `WindowManager.destroy()` be private? It doesn't fully destroy anymore, being run as one of the disposers, so it's really only for internal use now.

File: src/main/cluster-manager.ts [hunk 2]
```diff
@@ -68,25 +75,14 @@ export class ClusterManager extends Singleton {
...
-        entity.spec.metrics ||= { source: "local" };
-
-        if (entity.spec.metrics.source === "local") {
-          const prometheus: KubernetesClusterPrometheusMetrics = entity.spec?.metrics?.prometheus || {};
-
-          prometheus.type = cluster.preferences.prometheusProvider?.type;
-          prometheus.address = cluster.preferences.prometheus;
-          entity.spec.metrics.prometheus = prometheus;
-        }
-
```
Comment (jim-docker) [inline src/main/cluster-manager.ts]: why is this removed? Just a merge thing?

File: src/main/exit-app.ts [hunk 3]
```diff
@@ -26,19 +26,11 @@ import { ClusterManager } from "./cluster-manager";
...
-  console.log("before clusterManager");
-  const clusterManager = ClusterManager.getInstance(false);
-
-  console.log("after clusterManager");
-
   appEventBus.emit({ name: "service", action: "close" });
-  windowManager?.hide();
-  clusterManager?.stop();
+
+  WindowManager.resetInstance();
```
Comment (jim-docker) [inline src/main/exit-app.ts:31]: should hide() still be called, or is it moot since we're exiting?

**Review by jim-docker — COMMENTED** [review]
I see this after quitting Lens (run from `make dev`):
Screen Shot 2021-05-19 at 12 35 36 PM
It took about a minute to appear for me.
```[0] info: APP:QUIT
[0] info: APP:QUIT
[0] objc[94246]: _NSLocalEventObserver object 0x7fb20bb84c10 overreleased while already deallocating; break on objc_overrelease_during_dealloc_error to debug
[0] objc[94246]: _NSLocalEventObserver object 0x7fb20bb84c10 overreleased while already deallocating; break on objc_overrelease_during_dealloc_error to debug
[0] objc[94246]: _NSLocalEventObserver object 0x7fb20bb84c10 overreleased while already deallocating; break on objc_overrelease_during_dealloc_error to debug
[0] info: APP:QUIT
[0] info: APP:QUIT
[0] info: APP:QUIT
[0] objc[94246]: _NSLocalEventObserver object 0x7fb20bb84c10 overreleased while already deallocating; break on objc_overrelease_during_dealloc_error to debug
[0] objc[94246]: _NSLocalEventObserver object 0x7fb20bb84c10 overreleased while already deallocating; break on objc_overrelease_during_dealloc_error to debug
[0] info: APP:QUIT
[0] [nodemon] clean exit - waiting for changes before restart
```

File: src/main/cluster-manager.ts (hunk 2 above)
Comment (Nokel81) [reply src/main/cluster-manager.ts]: Oops yes it is a merge thing.

File: src/main/exit-app.ts (hunk 3 above)
Comment (Nokel81) [reply src/main/exit-app.ts:31]: I would say moot

**Comment by jim-docker** [comment]
looking good but this is still happening (when quitting installed OpenLens):

Screen Shot 2021-05-20 at 4 16 34 PM

**Comment by Nokel81** [comment]
Thanks will take a look

**Comment by Nokel81** [comment]
Closing this as this is a bad design.
